digitalocean/datasource_digitalocean_regions.go

```diff
@@ -0,0 +1,86 @@
+package digitalocean
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/digitalocean/godo"
+	"github.com/hashicorp/terraform/helper/hashcode"
+	"github.com/hashicorp/terraform/helper/schema"
+)
+
+func dataSourceDigitalOceanRegions() *schema.Resource {
+	return &schema.Resource{
+		Read: dataSourceDigitalOceanRegionsRead,
+		Schema: map[string]*schema.Schema{
+			"regions": {
+				Type:     schema.TypeSet,
+				Computed: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"name": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+						"slug": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+						"sizes": {
+							Type:     schema.TypeSet,
+							Elem:     &schema.Schema{Type: schema.TypeString},
+							Computed: true,
+						},
+						"features": {
+							Type:     schema.TypeSet,
+							Elem:     &schema.Schema{Type: schema.TypeString},
+							Computed: true,
+						},
+						"available": {
+							Type:     schema.TypeBool,
+							Computed: true,
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+func dataSourceDigitalOceanRegionsRead(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*CombinedConfig).godoClient()
+
+	opts := &godo.ListOptions{
+		Page:    1,
+		PerPage: 200,
+	}
+
+	regionList := []godo.Region{}
+	for {
+		regions, resp, err := client.Regions.List(context.Background(), opts)
+
+		if err != nil {
+			return fmt.Errorf("Error retrieving ssh keys: %s", err)
+		}
+
+		for _, region := range regions {
+			regionList = append(regionList, region)
+		}
+
+		if resp.Links == nil || resp.Links.IsLastPage() {
+			break
+		}
+
+		page, err := resp.Links.CurrentPage()
+		if err != nil {
+			return fmt.Errorf("Error retrieving ssh keys: %s", err)
+		}
+
+		opts.Page = page + 1
+	}
+
+	d.SetId(fmt.Sprintf("%d", hashcode.String(fmt.Sprintf("%+v\n", opts))))
+	d.Set("regions", regionList)
```

Comment:
For this kind of nested schema, you'll need to loop through the results with something like:

```
	regionsResult := make([]map[string]interface{}, 0)
	for _, r := range regionList {
		item := make(map[string]interface{})
		item["name"] = r.Name
		item["slug"] = r.Slug
		regionsResult = append(regionsResult, item)
	}

	d.Set("regions", regionsResult)
```

Example elsewhere in the code base:
https://github.com/terraform-providers/terraform-provider-digitalocean/blob/121c319d89b21926cb92fe1557c6c88f5e4294ce/digitalocean/resource_digitalocean_database_cluster.go#L334